sdk/dotnet/Pulumi/Core/Input.cs

```diff
@@ -21,29 +23,34 @@ internal interface IInput
     /// </summary>
     public class Input<T> : IInput
     {
-        /// <summary>
-        /// Technically, in .net we can represent Inputs entirely using the Output type (since
-        /// Outputs can wrap values and promises).  However, it would look very weird to state that
-        /// the inputs to a resource *had* to be Outputs. So we basically just come up with this
-        /// wrapper type so things look sensible, even though under the covers we implement things
-        /// using the exact same type
-        /// </summary>
-        private protected Output<T> _outputValue;
+        private readonly OneOf<T, Output<T>> _value;
+
+        private protected Input(T value)
+            => _value = OneOf<T, Output<T>>.FromT0(value);
 
         private protected Input(Output<T> outputValue)
-            => _outputValue = outputValue ?? throw new ArgumentNullException(nameof(outputValue));
+            => _value = OneOf<T, Output<T>>.FromT1(outputValue ?? throw new ArgumentNullException(nameof(outputValue)));
+
+        private protected virtual Output<T> ToOutput()
+            => _value.Match(v => Output.Create(v), v => v);
+
+        private protected virtual object? Value
+            => _value.Value;
```

Comment:
Because virtual, it's no longer true the type of this is OneOf<T, Output<T>>, it can be something else, which is taken advantage of in InputList..

Reply:
I think it might be worth annotating this prop with a note that this object is how serializer sees it, and that it can be different from  `OneOf<T, Output<T>>` introducing transforms. I also wonder what about deserializer. The hook can break symmetry e.g. can InputMap deserialize from what it serializes to? Or are we riding on the understanding that InputMaps never deserialize? Where you might be doing this is for example in writing typed provider mocks (not a feature yet), but when you're mocking a function F you might want to write your mock in terms of typed F.args, but mock framework would need to reconstruct F.args from the serial form.

Reply:
> Because virtual, it's no longer true the type of this is OneOf<T, Output>, it can be something else, which is taken advantage of in InputList

Right.

> I think it might be worth annotating this prop with a note

Yes, sounds good!

> I also wonder what about deserializer. The hook can break symmetry e.g. can InputMap deserialize from what it serializes to? Or are we riding on the understanding that InputMaps never deserialize?

This is a good question. I don't think the InputMap is going to show up during deserialization. The deserializer deserializes protobuf structs as `ImmutableDictionary<string, object?>`:

https://github.com/pulumi/pulumi/blob/f98d39b84bc0e3a78a13dc09f80730e39a297b10/sdk/dotnet/Pulumi/Serialization/Deserializer.cs#L75-L100

Typically, when you have an input typed as `InputMap<string>` the output will be `Output<ImmutableDictionary<string, string>>`.

> Where you might be doing this is for example in writing typed provider mocks (not a feature yet)

Hmmm. This is a good question. This shouldn't be a problem with our current `IMocks` abstraction, but typed provider mocks is interesting to think about. This will also likely come up when we get around to implementing nice provider APIs for .NET (for multi-lang components) where we'll want to take serialized inputs and deserialize them into the inputs that a component expects, which could be `InputList<T>` or `InputMap<V>`. Let me think about this some more. I suspect we'll be able to make this work without too much trouble.

Reply:
Thanks. My general intuition (from before Pulumi) is that serializers always end up needing to be symmetrical after more and more features pile up. Curious where we land on this.